### PR TITLE
CKAN 2.10

### PR DESCRIPTION
**Pull request by nickumia-reisys**
Related to
- https://github.com/GSA/data.gov/issues/4209
- https://github.com/GSA/data.gov/issues/4353
- https://github.com/GSA/ckanext-datajson/pull/144
- https://github.com/GSA/ckanext-googleanalyticsbasic/pull/15
- https://github.com/GSA/ckanext-usmetadata/pull/70
- https://github.com/GSA/ckanext-dcat_usmetadata/pull/308
  - https://github.com/GSA/ckanext-dcat_usmetadata/pull/311
- `ckanext-xloader` main was `CKAN 2.10` ready
- `ckanext-saml2auth` main was `CKAN 2.10` ready
- `ckanext-s3filestore` main was `CKAN 2.10` ready
- `ckanext-envvars == 0.0.3` is `CKAN 2.10` ready

Changes:
- Upgrade all core requirements to CKAN 2.10 minimums
- Switch test docker image from `2.9` to `2.10`
- (BREAKING?) Fix `datagov_disallow_anonymous_access` function
- (BREAKING) `h.flash.pop_messages` is deprecated and replaced with `h.get_flashed_messages`
- Update CKAN extension version check (cypress)
- Update `flHideToolBarButton` to `flDebugHideToolBarButton` (cypress)
- Add `{force: true}` to resource download test (cypress)
- Fix failed login validation (cypress)
- Setup `ckan.datapusher.api_token` because of defaults in CKAN 2.10 that require it (I assume it's only for the test image and not for production)

### CODE REVIEW

**Comment by nickumia-reisys** [comment]
Error deploying to development...
```bash
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR Traceback (most recent call last):
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/bin/ckan", line 8, in <module>
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR sys.exit(ckan())
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR return self.main(*args, **kwargs)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/click/core.py", line 1054, in main
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR with self.make_context(prog_name, args, **extra) as ctx:
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/click/core.py", line 920, in make_context
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR self.parse_args(ctx, args)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/ckan/cli/cli.py", line 121, in parse_args
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR result = super().parse_args(ctx, args)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/click/core.py", line 1613, in parse_args
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR rest = super().parse_args(ctx, args)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/click/core.py", line 1378, in parse_args
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR value, args = param.handle_parse_result(ctx, opts, args)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/click/core.py", line 2360, in handle_parse_result
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR value = self.process_value(ctx, value)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/click/core.py", line 2322, in process_value
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR value = self.callback(ctx, self, value)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/ckan/cli/cli.py", line 131, in _init_ckan_config
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR _add_ctx_object(ctx, value)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/ckan/cli/cli.py", line 140, in _add_ctx_object
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR ctx.obj = CtxObject(path)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/ckan/cli/cli.py", line 57, in __init__
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR self.app = make_app(raw_config)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/ckan/config/middleware/__init__.py", line 29, in make_app
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR flask_app = make_flask_stack(conf)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/ckan/config/middleware/flask_app.py", line 275, in make_flask_stack
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR log.warn(csrf_warn_extensions)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/logging/__init__.py", line 1463, in warn
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR self.warning(msg, *args, **kwargs)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/logging/__init__.py", line 1458, in warning
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR self._log(WARNING, msg, args, **kwargs)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/logging/__init__.py", line 1589, in _log
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR self.handle(record)
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/logging/__init__.py", line 1598, in handle
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR if (not self.disabled) and self.filter(record):
   2023-06-05T15:37:19.00-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/logging/__init__.py", line 806, in filter
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR result = f.filter(record)
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/ckan/config/middleware/flask_app.py", line 643, in filter
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR log_record.url = request.path
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/werkzeug/local.py", line 436, in __get__
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR obj = instance._get_current_object()
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/werkzeug/local.py", line 565, in _get_current_object
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR return self.__local()  # type: ignore
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.9/site-packages/flask/globals.py", line 33, in _lookup_req_object
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR raise RuntimeError(_request_ctx_err_msg)
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR RuntimeError: Working outside of request context.
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR This typically means that you attempted to use functionality that needed
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR an active HTTP request.  Consult the documentation on testing for
   2023-06-05T15:37:19.01-0400 [APP/PROC/WEB/0] ERR information about how to avoid this problem.
```

File: ckanext/datagov_inventory/tests/logic/auth/test_auth.py [hunk 1]
```diff
@@ -371,28 +371,40 @@ def test_auth_task_status_show(self):
...
             'anonymous': is_denied
         })
 
     def test_auth_user_show(self):
+        '''
+        IMPORTANT CHANGE: This test allowed all users access to the 'user_show'
+        function until CKAN 2.10.  We don't know why the functionality
+        changed.  But only sysadmin can access this function in tests.
+        From the UI, it seems like all users have access to this operation.
+        '''
```
Comment (nickumia-reisys) [inline ckanext/datagov_inventory/tests/logic/auth/test_auth.py:398]: Important references:
- https://github.com/ckan/ckan/blob/ckan-2.10.0/ckan/tests/helpers.py#L85
- https://github.com/ckan/ckan/blob/ckan-2.10.0/ckan/logic/auth/__init__.py#L93
- https://github.com/ckan/ckan/blame/ckan-2.10.0/ckan/logic/action/get.py#LL1423C5-L1423C14
- https://github.com/ckan/ckan/blame/ckan-2.10.0/ckan/logic/auth/get.py#L200
- https://docs.ckan.org/en/2.10/maintaining/configuration.html#ckan-auth-public-user-details
- https://docs.ckan.org/en/2.10/api/index.html#ckan.logic.action.get.user_show

**Comment by nickumia-reisys** [comment]
For the issues deploying to Cloud.gov, we've isolated that it is either a ckan command or the `datastore-setup.py` that is causing the error.  If it is the `datastore-setup.py`, we need to figure out a way to do that within the flask web app context...

**Comment by nickumia-reisys** [comment]
New revelation: the app is failing on cloud.gov because of the database needs to be updated... if we run `ckan db upgrade`, it gives the `RuntimeError: Working outside of request context.` error. @FuhuXia

**Comment by btylerburton** [comment]
I was able to publish a new dataset on `develop` using the metadata-form based on this branch `ckan-210-startup-issues__tyler`

https://inventory-dev-datagov.app.cloud.gov/dataset/new-test

**Comment by btylerburton** [comment]
Compare diff: https://github.com/GSA/inventory-app/compare/ckan-210...ckan-210-startup-issues__tyler

File: ckanext/datagov_inventory/plugin.py
Comment (nickumia-reisys) [inline ckanext/datagov_inventory/plugin.py:1]: References for creating the custom blueprint redirect:
- https://github.com/search?q=repo%3Ackan%2Fckan+%22render%22&type=code
- https://github.com/search?q=repo%3Ackan%2Fckan+%22search_template%22&type=code
- https://github.com/search?q=repo%3Ackan%2Fckan+%22index_template%22&type=code
- https://github.com/ckan/ckan/blob/dd4f061c4a521817bd796e09b78b613aaf1a0123/ckan/plugins/interfaces.py#L1409
- https://github.com/ckan/ckan/blob/dd4f061c4a521817bd796e09b78b613aaf1a0123/ckan/views/group.py#L184
- https://github.com/ckan/ckan/blob/dd4f061c4a521817bd796e09b78b613aaf1a0123/ckan/lib/plugins.py#L565
- https://github.com/ckan/ckan/blob/dd4f061c4a521817bd796e09b78b613aaf1a0123/ckan/views/__init__.py#L11
- https://github.com/ckan/ckan/blob/dd4f061c4a521817bd796e09b78b613aaf1a0123/ckan/views/home.py#L15
- https://github.com/GSA/ckanext-geodatagov/blob/main/ckanext/geodatagov/blueprint.py
- https://github.com/GSA/ckanext-datagovtheme/blob/main/ckanext/datagovtheme/plugin.py
- https://github.com/GSA/ckanext-datagovtheme/blob/main/ckanext/datagovtheme/blueprint.py

**Review by jbrown-xentity — COMMENTED** [review]
A few QA things:
1. The title site says "CKAN", that should say "Inventory"
2. Resource is not downloadable
3. When creating/editing metadata as an "editor" role, you get failure with CSRF on some API routes: https://inventory-dev-datagov.app.cloud.gov/api/3/action/organization_list_for_user

**Comment by jbrown-xentity** [comment]
> 3. When creating/editing metadata as an "editor" role, you get failure with CSRF on some API routes: https://inventory-dev-datagov.app.cloud.gov/api/3/action/organization_list_for_user

In order to validate this, we want an "editor" user implemented in our tests. In order to do this, we'll probably need to 
1. Create the user via UI (as sysadmin) (Go to organizations page, click admin, Create new user using form with editor permissions).
  2. You may want to do this by re-creating the sysadmin account to extract the api key, and then utilize the api to create the user with the right permissions.
3. Add user to organization (if applicable)
4. Logout as sysadmin
5. Login as new user
6. Then test dataset creation

File: config/ckan.ini
Comment (nickumia-reisys) [inline config/ckan.ini:1]: For future reference: the inventory-logo lives here: https://github.com/GSA/ckanext-usmetadata/tree/main/ckanext/usmetadata/public/base/images

File: requirements.in.txt [hunk 2]
```diff
@@ -23,46 +27,43 @@ urllib3==1.26.5
...
-pyyaml==5.4
-repoze.who==2.3
-requests==2.31.0
-Routes==1.13
-rq==1.0
+pysolr==3.9.0
+pytz
+PyUtilib==6.0.0
+pyyaml==6.0
+requests==2.28.1
```
Comment (nickumia-reisys) [inline requirements.in.txt:56]:
```suggestion
requests==2.31.0
```

Took the wrong version of `requests`

File: requirements.txt [hunk 3]
```diff
@@ -50,55 +53,55 @@ Mako==1.2.4
...
-repoze.lru==0.7
-repoze.who==2.3
-requests==2.31.0
+python-magic==0.4.27
+pytz==2023.3
+pytz-deprecation-shim==0.1.0.post0
+PyUtilib==6.0.0
+PyYAML==6.0
+redis==4.5.5
+requests==2.28.1
```
Comment (nickumia-reisys) [inline requirements.txt:79]:
```suggestion
requests==2.31.0
```
Took the wrong version of `requests` (different file)

**Comment by nickumia-reisys** [comment]
> A few QA things:
> 
>     1. The title site says "CKAN", that should say "Inventory"
> 
>     2. Resource is not downloadable
> 
>     3. When creating/editing metadata as an "editor" role, you get failure with CSRF on some API routes: https://inventory-dev-datagov.app.cloud.gov/api/3/action/organization_list_for_user

All three of these issues should be resolved
(1.) Was never broken on cloud.gov.  It just wasn't configured for a local instance
  - https://github.com/GSA/inventory-app/pull/611/commits/bc13724638d9a77581339ff10d8bda2265a7db35
(2.) Was fixed in `ckanext-usmetadata`
  - https://github.com/GSA/ckanext-usmetadata/pull/70/commits/dbadd6335fe1f1aa318bdf21e23b821cae8862e4
  - https://github.com/GSA/ckanext-usmetadata/pull/70/commits/40788f78603fbc9386ab883e0cc48abaa82c4036
(3.) Was fixed in `ckanext-dcat_usmetadata` (still needs to be integrated though)
  - https://github.com/GSA/ckanext-dcat_usmetadata/pull/311

**Comment by nickumia-reisys** [comment]
@GSA/data-gov-team 

⚠️ DO NOT MERGE THIS PR UNTIL VERIFYING WITH @nickumia-reisys . ⚠️

**Comment by jbrown-xentity** [comment]
I still get the CSRF error in staging, with an admin account... Is my account somehow just jacked up? Can anyone else replicate this error?
image

**Comment by nickumia-reisys** [comment]
Inventory staging was running an older versoin of `ckanext-dcat_usmetadata`.  After deploying the latest version of the extension, the issue @jbrown-xentity's found was fixed :)

**Review by Jin-Sun-tts — APPROVED** [review]

**Pull request merged**